Fix rendering of voxels on top/bottom of screen

MrAlaux opened the pull request:
Rendering is really not my field and I hardly know any math, so I may be understanding this very wrongly, but I've spent hours looking at the voxel code trying to figure out what causes #1277 and I think I got it.

When we enter the slab loop in `VX_DrawColumn()` (a "slab" is apparently a cube in a column of the voxel model, to put it simply), we calculate `uy1` and `uy2`, which seem to be the vertical coordinates between which the current pixel column of the side of the slab should be drawn (i.e. we draw a column of pixels from `uy1` down to `uy2)`. The top of the slab is drawn from slightly above `uy1` down to `uy1` itself, whereas the bottom of the slab is drawn from slightly below `uy2` up to `uy2` itself.

The problem lies in the fact that when either `uy1` goes below the bottom edge of the screen, or `uy2` goes above the top edge, the whole slab is omitted. This makes sense for the sides of the slab, but not for its top or bottom, which as I said slightly exceed `uy1` and `uy2`.

The little change I've made in this PR makes the loop still omit drawing of sides in that case, but it keeps going to see if the top or bottom should be drawn; it fixes #1277.

@andwj thoughts?

fabiangreffrath commented:
Apparently that fixed it, good job!

MrAlaux commented:
> Apparently that fixed it, good job!

Yes, but I'm afraid that something could be wrong; I did trigger two or three crashes just looking around a voxel model, but couldn't reproduce them later. I'd like some reviews.

andwj commented:
I have re-acquainted myself with the code, and I'm fairly sure this change is okay - e.g. will not draw outside of the clipping range.

MrAlaux commented:
> I have re-acquainted myself with the code, and I'm fairly sure this change is okay - e.g. will not draw outside of the clipping range.

Thanks for checking!

Regarding the crashes I reported, I'm guessing that the only other possible culprit would be the `wscale` calculation; in cases where `(has_top || has_bottom) && !has_side`, is it possible that it could lead to a division by zero?

fabiangreffrath commented:
> Regarding the crashes I reported, I'm guessing that the only other possible culprit would be the `wscale` calculation; in cases where `(has_top || has_bottom) && !has_side`, is it possible that it could lead to a division by zero?

I'd like to investigate this. Do you have any demo that crashes reliably? Is this resolution dependent?

MrAlaux commented:
> Do you have any demo that crashes reliably? Is this resolution dependent?

I did try a few demos with different settings, but got no crashes -- no idea if it depends on resolution either.

However, like I mentioned in the Nugget Discord, I had the idea of tweaking the demotest to reenable rendering and load voxels, and also to disable `-fastdemo` to get more frames rendered. I haven't done that yet, so feel free to do it yourself.

rfomin commented:
I ran some long demos with the ASAN build - no crashes so far.

> Regarding the crashes I reported, I'm guessing that the only other possible culprit would be the `wscale` calculation; in cases where `(has_top || has_bottom) && !has_side`, is it possible that it could lead to a division by zero?

Why? I don't see division by `wscale`, only multiplication.

fabiangreffrath commented:
I have also yet to reproduce a crash with this change.

MrAlaux commented:
> Why? I don't see division by `wscale`, only multiplication.

I was talking about the calculations whose values are assigned to `wscale`, which include `FixedDiv()`'s; perhaps some of the cases where the now-fixed bug was occurring also happened to be cases where the divisor ended up being zero.

It might be possible that I added some test code locally to check if `I_VideoBuffer[]` was being exceeded (i.e. columns being drawn out of the clipping range), and that was actually what caused the crashes, though I doubt it. Either way, if any further testing doesn't provoke any crashes, I guess we could merge this as it is and revert it if we get any crash reports.

MrAlaux commented:
FWIW, I've just merged this fix into Nugget, and I plan to make a release soon following today's new Woof release; that could help to determine if there really is a crash provoked by the fix.

rfomin commented:
> FWIW, I've just merged this fix into Nugget, and I plan to make a release soon following today's new Woof release; that could help to determine if there really is a crash provoked by the fix.

There seem to be no issues? Should we merge this?

fabiangreffrath commented:
I'd say we should merge this.

MrAlaux commented:
Fine by me too. However, seeing as this is a performance-heavy function, should we apply the following?

```diff
diff --git a/src/r_voxel.c b/src/r_voxel.c
index 9cdf9d29..715d5d50 100644
--- a/src/r_voxel.c
+++ b/src/r_voxel.c
@@ -870,15 +870,29 @@ static void VX_DrawColumn (vissprite_t * spr, int x, int y)
 			fixed_t uy2 = uy1 + (fixed_t) len * scale;
 			fixed_t uy0 = uy1;
 
+			boolean has_side = ((face & (ux > Bx ? B_face : A_face)) != 0);
+
 			// clip the slab vertically
-			if (uy1 >= clip_y2) uy1 = clip_y2;
-			if (uy2 <= clip_y1) uy2 = clip_y1;
 
-			if (uy1 < clip_y1) uy1 = clip_y1;
-			if (uy2 > clip_y2) uy2 = clip_y2;
+			if (uy1 < clip_y1)
+			{
+				uy1 = clip_y1;
+			}
+			else if (uy1 >= clip_y2)
+			{
+				uy1 = clip_y2;
+				has_side = false;
+			}
 
-			boolean has_side = ((face & (ux > Bx ? B_face : A_face)) != 0
-                          && uy1 < clip_y2 && uy2 > clip_y1);
+			if (uy2 > clip_y2)
+			{
+				uy2 = clip_y2;
+			}
+			else if (uy2 <= clip_y1)
+			{
+				uy2 = clip_y1;
+				has_side = false;
+			}
 
 			// handle the fuzz effect for Spectres
 			if (shadow)
@@ -917,8 +931,14 @@ static void VX_DrawColumn (vissprite_t * spr, int x, int y)
 
 				uy = ((uy - 1) | FRACMASK) + 1;
 
-				if (uy < clip_y1)
+				if (uy >= clip_y2)
+				{
+					break;
+				}
+				else if (uy < clip_y1)
+				{
 					uy = clip_y1;
+				}
 
 				byte src = slab[0];
 				byte pix = spr->colormap[spr->brightmap[src]][src];
@@ -932,8 +952,14 @@ static void VX_DrawColumn (vissprite_t * spr, int x, int y)
 			{
 				fixed_t uy = centeryfrac - FixedMul (top_z - ((int)len << FRACBITS), wscale);
 
-				if (uy > clip_y2)
+				if (uy <= clip_y1)
+				{
+					continue;
+				}
+				else if (uy > clip_y2)
+				{
 					uy = clip_y2;
+				}
 
 				byte src = slab[len - 1];
 				byte pix = spr->colormap[spr->brightmap[src]][src];

```

rfomin commented:
> Fine by me too. However, seeing as this is a performance-heavy function, should we apply the following?

Looks reasonable, but I doubt there will be a measurable difference. Maybe we should keep the changes to a minimum.

fabiangreffrath commented:
> Looks reasonable, but I doubt there will be a measurable difference. Maybe we should keep the changes to a minimum.

Agreed. We should stay as close to the original code as possible and leave such optimizations to the compiler.
